apps/meteor/ee/server/apps/marketplace/fetchMarketplaceApps.ts

```diff
@@ -0,0 +1,177 @@
+		bundledIn: v
+			.array(
+				v.object({
+					bundleId: v.string(),
+					bundleName: v.string(),
+					addonTierId: v.string(),
+				}),
+			)
+			.required(),
+		isPurchased: v.boolean().required(),
+		isSubscribed: v.boolean().required(),
+		subscriptionInfo: v.object({
+			typeOf: v.string().enum('app', 'service').required(),
+			status: v.string().enum('trialing', 'active', 'cancelled', 'cancelling', 'pastDue').required(),
+			statusFromBilling: v.boolean().required(),
+			isSeatBased: v.boolean().required(),
+			seats: v.number().required(),
+			maxSeats: v.number().required(),
+			license: v
+				.object({
+					license: v.string().required(),
+					version: v.number().required(),
+					expireDate: v.string().required(),
+				})
+				.required(),
+			startDate: v.string().required(),
+			periodEnd: v.string().required(),
+			endDate: v.string(),
+			externallyManaged: v.boolean().required(),
+			isSubscribedViaBundle: v.boolean().required(),
+		}),
+		price: v.number().required(),
+		purchaseType: v.string().enum('', 'buy', 'subscription').required(),
+		pricingPlans: v.array(
+			v.object({
+				id: v.string().required(),
+				enabled: v.boolean().required(),
+				price: v.number().required(),
+				trialDays: v.number().required(),
+				strategy: v.string().enum('once', 'monthly', 'yearly').required(),
+				isPerSeat: v.boolean().required(),
+				tiers: v.array(
+					v.object({
+						perUnit: v.boolean().required(),
+						minimum: v.number().required(),
+						maximum: v.number().required(),
+						price: v.number().required(),
+						refId: v.string(),
+					}),
+				),
+			}),
+		),
+		isUsageBased: v.boolean().required(),
+
+		requestedEndUser: v.boolean(),
+		requested: v.boolean(),
+		appRequestStats: v.object({}),
+
+		createdAt: v.string().required(),
+		modifiedAt: v.string().required(),
+	}),
+);
+
+const assertWorkspaceCommPayload = compile(fetchMarketplaceAppsSchema);
+
+export async function fetchMarketplaceApps({ endUserID }: FetchMarketplaceAppsParams = {}): Promise<App[]> {
+	const baseUrl = Apps.getMarketplaceUrl();
+	const headers = getMarketplaceHeaders();
+	const token = await getWorkspaceAccessToken();
+	if (token) {
+		headers.Authorization = `Bearer ${token}`;
+	}
+
+	let request;
+	try {
+		request = await fetch(`${baseUrl}/v1/apps`, {
+			headers,
+			params: {
+				...(endUserID && { endUserID }),
+			},
+		});
+	} catch (error) {
+		throw new MarketplaceConnectionError('Marketplace_Bad_Marketplace_Connection');
+	}
+
+	if (request.status === 200) {
+		const response = await request.json();
+		assertWorkspaceCommPayload(response);
+		return response;
+	}
+
+	const response = await request.json();
+	if (request.status === 400 && response.code === 200) {
+		throw new MarketplaceAppsError('Marketplace_Invalid_Apps_Engine_Version');
+	}
+
+	const INTERNAL_MARKETPLACE_ERROR_CODES = [266, 256, 166, 221, 257, 320];
+
+	if (request.status === 500 && INTERNAL_MARKETPLACE_ERROR_CODES.includes(response.code)) {
```

Comment:
Should we at least debuglog the actual error code/message?

Reply:
oh yea, sure thing